### ISSUE TITLE
External layouts

### ISSUE THREAD

**Issue by Leon-Plickat** [ifreund/river#28]
As discussed in IRC, there should be a way to let external executables generate a layout. I opened this Issue as a place to brainstorm this.

Communication between river and layout generator aside, there needs to be a sane way to tell river to try get a layout from an external program. My idea would be to overload the `layout` command. I believe this could be done by modifying `Output.getLayoutByName()` to check whether the provided string is a path to an executable file (after checking for the names of the build-in layouts). Also needed would be another member of the `Output.Layout` enum (maybe `External`), a new variable for the `Output` class storing the path and a new layout function (`Output.layoutExternal()` ) for that class, that abstracts talking to the executable. This is the simplest and most painless way to integrate external layouts I could come up with.

Of interest is also what happens in `Output.layoutExternal()`: Should it handle padding, borders and offsets, so that the layout generator can be as simple as possible, or should we omit that to give more power to users wanting to write obscure/complicated layouts? And what data should the generator be provided with? Should a simple use case like this really use JSON, if it would also be sufficient to just read space separated numbers? And wouldn't it make more sense to provide the data to the generator as arguments, rather than piping that into it via `stdin`, so that the generators code could be less complex?

**Comment by ifreund** [comment]
I think my ideal would be making this feature powerful enough that we can drop builtin layouts completely and just ship a few basic ones as separate binaries. Then the `layout` command wouldn't need to be overloaded and would just accept a path to the external executable (or something in the $PATH). 

I'm of the opinion that the compositor should handle all padding/borders/offsets internally as they are really orthogonal to the actual implementation of the layout. I could potentially be convinced otherwise given a concrete example of something worthwhile that this additional complexity would make possible.

I mentioned JSON as a possibility as nearly every language has a library for reading/writing it and it provides structure to the input/output. Perhaps passing the data as arguments would simplify things though. Looking at what we'd actually need to pass to the "generator" that may make more sense as it's really not too much unless I'm missing something:

Sample input (could easily just be 5 arguments as well)
```json
{
    "view_count": 4,
    "master_count": 1,
    "master_factor": 0.6,
    "width": 500,
    "height": 400
}
```
Sample output (I think json makes more sense here):
```json
[
    {"x":  0, "y": 	0, "width": 42, "height": 42},
    {"x": 42, "y": 0, "width": 42, "height": 42},
    {"x": 84, "y": 0, "width": 42, "height": 42},
    {"x": 0,  "y": 42, "width": 42, "height": 42}
]
```

**Comment by Leon-Plickat** [comment]
I think at least one internal layout should be implemented as a fallback. When river starts up, the entire compositor startup is done before the users config script can set the layout and it is possible that windows are spawned before that. Its the same when a new output is added. I am personally fine with leaving the ones we already have (master-stack and full) in the code, but if you want to simplify it, the only internal layout could be a dummy layout that just places every window at `0,0` and does not change their size  (so that the windows preferred size is used).

You are right about JSON output (especially since it is relatively easy to produce even if the language used does not have a library), but for input I'd prefer arguments. And for the sake of simplicity I'll agree with letting river handle borders and padding internally.

**Comment by ifreund** [comment]
True, we probably do need to keep some kind of hardcoded layout in river. I think the decision on what exactly that looks like can wait until we have external layouts implemented and see how that works out.

I agree, arguments make more sense as there should always be exactly 5 values passed in the same order. Would certainly help simplify the posix shell generators that we both know are coming :D

**Comment by Leon-Plickat** [comment]
> Would certainly help simplify the posix shell generators that we both know are coming :D 

I tried it and its easier than anticipated. Simple full layout in shell:

```sh
#!/bin/sh

CLIENTS="$1"
WIDTH="$4"
HEIGHT="$5"

printf "[\n"

for i in $(seq 1 "$CLIENTS")
do
    if [ "$i" = "$CLIENTS" ]
    then
         COMMA=""
    else
         COMMA=","
    fi
    printf "\t{\"x\": %d, \"y\": %d, \"width\": %d, \"height\": %d}%c\n" 0 0  "$WIDTH" "$HEIGHT" "$COMMA"
done

printf "]\n"
```

Admittedly its cheating since nothing needs to be calculated. So for complex layouts you might need bash instead of pure POSIX shell.

**Comment by ifreund** [comment]
After thinking about this some more I came to the conclusion that it would simplify things if generators simply output the 4 numbers per line in the right order. You're right, JSON really doesn't win us much here.

**Comment by Leon-Plickat** [comment]
Yeah, I have done some testing (nothing PR worthy yet). Using JSON would mean to either add another dependency (like json-c) and integrate another C library, or write our own minimal JSON handler. Both approaches were a bit to much for the time I have available right now.

But parsing four values with a newline delimiter should be ridiculously easy; I'll have another look at it.

**Comment by ifreund** [comment]
Cool, if you run into any issues let me know. There is actually a json implementation in the zig standard library currently which I was planning on using, but it will keep everything much simpler to just not mess with it.

**Comment by Leon-Plickat** [comment]
Closed via #48
